site/content/kapp-controller/docs/latest/private-registry-auth.md

```diff
@@ -0,0 +1,102 @@
+---
+title: Authenticating to Private Registires
+---
+
+In order for kapp-controller to pull imgpkg bundles or images from registries requiring authentication, 
+kapp-controller supports a workflow to more easily manage Kubernetes secrets and use them as part of working 
+with PackageRepositories or PackageInstalls. 
+
+In addition to having kapp-controller installed, when [secretgen-controller](https://github.com/vmware-tanzu/carvel-secretgen-controller) 
+is installed on the same cluster, kapp-controller will create secrets that are used when PackageRepositories or 
+PackageInstalls are created. These secrets are then populated by secretgen-controller and will contain all "exported" 
+registry credentials stored on a cluster as Kubernetes secrets. 
+
+By simply adding the credentials as Kubernetes secrets in the flows mentioned below, kapp-controller will be able 
+to successfully authenticate to registries without needing to manage secret references in PackageRepositories 
+or PackageInstalls.
+
+### Installing secretgen-controller
+
+To use the authentication flows mentioned below, install secretgen-controller on the cluster where kapp-controller is installed:
+
+```bash
+kapp deploy -a sg -f https://raw.githubusercontent.com/vmware-tanzu/carvel-secretgen-controller/develop/alpha-releases/0.4.0-alpha.1.yml
+```
+
+### Create Registry Credentials
+
+**NOTE:** Currently the only secret type supported for this workflow is `type: kubernetes.io/dockerconfigjson`. All 
+secrets created for this approach should be in this format. More on the dockerconfigjson type can be found [here](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#registry-secret-existing-credentials).
+
+After installing secretgen-controller on the same cluster as where kapp-controller is installed, registry 
+credentials can be created as Kubernetes secrets and exported to specific namespaces on a cluster as shown below:
+
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: regcred
+  namespace: packagerepo-ns
+type: kubernetes.io/dockerconfigjson
+stringData:
+  .dockerconfigjson: |
+    {
+      "auths": {
+        "index.docker.io/v2/": {
```

Comment:
Found out that `index.docker.io/v2/` is going to be problematic. should be just `index.docker.io` since thats how imgpkg matches hostname. cc @DennisDenuto should we enhance imgpkg to do better matching for ..._HOSTNAME?

Reply:
Thanks, changed

Reply:
@cppforlife 
I'm curious what you found to be problematic with `index.docker.io/v2/`.
From what I can see having  `index.docker.io/v2/` in a dockerconfig.json would result in imgpkg keychain correctly resolving images belonging to the registry `index.docker.io`. We have a test for this case too https://github.com/vmware-tanzu/carvel-imgpkg/blob/48f3c82ce39d4b70099c8d4c627e523be6eed419/pkg/imgpkg/registry/keychain_test.go#L182-L188

> should we enhance imgpkg to do better matching for ..._HOSTNAME
I'm understanding this to mean, allow setting URL paths in the `REGISTRY_HOSTNAME` env variable?

Hmm theres a tradeoff here. between making it easier for folk to set this env variable (they don't have to strip out the URL path) vs potentially losing the strict validation checks catching user errors / typos.

for e.g. `REGISTRY_HOSTNAME=index.docker/io` would match `index.docker` as the registry hostname and probably fail with a http auth related error instead of the hostname validation error they get today. 

I haven't heard of any pain around imgpkg's strict _HOSTNAME validation logic yet (Unless y'all have - which, if you have, i'll submit a PR to change that validation logic to make it more permissive)